migrations/20231128_jellyfish_migration/utils/data_migration.go

```diff
@@ -0,0 +1,386 @@
+		updateOp.SetUpdate(bson.M{"$unset": bson.M{rollbackSectionName: ""}})
+	}
+	updates = append(updates, updateOp)
+	return updates
+}
+
+func (m *DataMigration) SetUpdates(data UpdateData) {
+	m.groupedDiffs[data.ItemType] = append(m.groupedDiffs[data.ItemType], data)
+	m.updates = append(m.updates, data.getMongoUpdates(m.settings.Rollback, m.settings.RollbackSectionName)...)
+}
+
+func (m *DataMigration) updatesApplied(updatedCount int) {
+	m.updates = []mongo.WriteModel{}
+	m.updatedCount += updatedCount
+}
+
+func (m *DataMigration) completeUpdates() bool {
+	capacity := m.settings.capacity
+	if capacity != nil {
+		stats := m.GetStats()
+
+		percent := (float64(stats.Fetched) * float64(100)) / float64(*capacity)
+
+		log.Printf("processed %.0f %% of %d records and applied %d changes", percent, *capacity, stats.Applied)
+
+		if *capacity <= stats.Applied || *capacity <= stats.Fetched {
+			log.Printf("cap [%d] updates applied [%d] fetched [%d]", *capacity, stats.Applied, stats.Fetched)
+			return true
+		}
+	}
+	return false
+}
+
+func (m *DataMigration) GetUpdates() []mongo.WriteModel {
+	return m.updates
+}
+
+func (m *DataMigration) SetLastProcessed(lastID string) {
+	m.lastUpdatedID = lastID
+	m.writeLastProcessed(m.lastUpdatedID)
+}
+
+func (m *DataMigration) SetFetched(raw []bson.M) {
+	m.fetchedCount += len(raw)
+}
+
+func (m *DataMigration) GetStats() MigrationStats {
+	return MigrationStats{
+		Errored: m.errorsCount,
+		Fetched: m.fetchedCount,
+		ToApply: len(m.updates),
+		Applied: m.updatedCount,
+		Elapsed: time.Since(m.startedAt).Truncate(time.Millisecond),
+	}
+}
+
+func (m *DataMigration) GetLastID() string {
+	return m.lastUpdatedID
+}
+
+func (m *DataMigration) OnError(data ErrorData) {
+	m.errorsCount++
+	m.groupedErrors[data.ItemType] = append(m.groupedErrors[data.ItemType], data)
+	if m.settings.StopOnErr {
+		log.Printf("[_id=%s] %s %s\n", data.ItemID, data.Msg, data.Error.Error())
+		os.Exit(1)
+	}
+}
+
+func (m *DataMigration) CheckMongoInstance() error {
+	if err := m.mongoInstanceChecker.BlockUntilDBReady(m.GetCtx()); err != nil {
+		return err
+	}
+	if err := m.mongoInstanceChecker.CheckFreeSpace(m.GetCtx(), m.GetDataCollection()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c MigrationStats) report() {
+	if c.Applied == 0 && c.Fetched > 0 {
+		log.Printf("elapsed [%s] for [%d] items fetched with [%d] errors\n", c.Elapsed, c.Fetched, c.Errored)
+		return
+	}
+	log.Printf("elapsed [%s] for [%d] fetched [%d] updates applied with [%d] errors\n", c.Elapsed, c.Fetched, c.Applied, c.Errored)
+}
+
+func (m *DataMigration) writeErrors(groupLimit *int) {
+	if !m.settings.writeToDisk {
+		m.groupedErrors = map[string][]ErrorData{}
+	}
+	for group, errors := range m.groupedErrors {
+		if groupLimit != nil {
+			if len(errors) < *groupLimit {
+				continue
+			}
+		}
+		f, err := m.createFile("error", group, "%s.log")
+		if err != nil {
+			log.Println(err)
+			os.Exit(1)
+		}
+		defer f.Close()
```

Comment:
This will leave all of the files open until the function exits. It won't close each file after it is written to (which I think was your intention).